### PR TITLE
Error attachment implemented as separate log

### PR DESCRIPTION
**Pull request by bmourat**
Error attachment implemented as ErrorAttachmentLog class which is send separatley from error log

### CODE REVIEW

File: apps/sasquatch/src/projectDependency/java/com/microsoft/azure/mobile/sasquatch/activities/MainActivity.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
+package com.microsoft.azure.mobile.sasquatch.activities;
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/azure/mobile/sasquatch/activities/MainActivity.java]: We should not fork the entire class, you could do it like this:
* Have ProjectCrashListener (in project source folder) extends the CrashListener that is common to both flavours.
* Define a tiny class with a static method that returns the crash listener (this class would be in 2 different flavours).
* Thus MainActivity remains in 1 flavour and calls the static method to get the listener.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/CrashesListener.java [hunk 2]
```diff
@@ -29,16 +30,15 @@
     @UiThread
     boolean shouldAwaitUserConfirmation();
 
-    /*
-     * Called to get additional information to be attached to a crash report before sending.
-     * Attachment is an optional so this method can also return <code>null</code>.
+    /**
+     * Called to get additional information to be send as separate ErrorAttachmentLog logs
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/CrashesListener.java]: send => sent

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java [hunk 3]
```diff
@@ -22,52 +25,87 @@
...
+     *
+     * @param log ErrorAttachmentLog to validate.
+     * @return true if validation succeeded, otherwise false.
+     */
+    static boolean validateErrorAttachmentLog(ErrorAttachmentLog log) {
+        if (log == null) {
+            MobileCenterLog.error(LOG_TAG, "Null ErrorAttachmentLog passed to validate method");
+            return false;
+        }
+        if( log.getId() == null || log.getErrorId() == null || log.getContentType() == null ||
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java]: file is not formatted, I can tell with the misplaced opening parenthesis ;)

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 4]
```diff
@@ -762,6 +759,24 @@ public void run() {
...
+        }
+        else {
+            for (ErrorAttachmentLog attachment : attachments) {
+                if(attachment != null) {
+                    attachment.setId(UUID.randomUUID());
+                    attachment.setErrorId(errorLogReport.log.getId());
+                    if(ErrorAttachments.validateErrorAttachmentLog(attachment))
+                        mChannel.enqueue(attachment, ERROR_GROUP);
+                } else {
+                    MobileCenterLog.error(LOG_TAG, "Skipping null ErrorAttachmentLog in CrashesListener.getErrorAttachments");
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: That should be a warning.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java [hunk 5]
```diff
@@ -22,52 +25,87 @@
...
+     * @param fileName    file name to use in error attachment log.
      * @param contentType binary data MIME type.
-     * @return error attachment or null if text and data are null.
+     * @return ErrorAttachmentLog attachment or null if text and data are null.
      */
-    public static ErrorAttachment attachment(String text, byte[] data, String fileName, String contentType) {
-        if (text == null && data == null) {
-            MobileCenterLog.warn(LOG_TAG, "Null content passed to attachment method, returning null");
+    private static ErrorAttachmentLog attachment(byte[] data, String text, String fileName, String contentType) {
+        if (contentType == null) {
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java]: Move validation to Crashes as discussed in the previous version of the PR, people could bypass the static method.

**Comment by jaelim-ms** [comment]
Please bring back TODO comments

**Comment by bmourat** [comment]
Brought back TODO comments. But why do we need them in this branch?

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java (hunk 5 above)
Comment (bmourat) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java]: I moved validatation to Crashes as we discussed.
https://github.com/Microsoft/mobile-center-sdk-android/pull/375/files#diff-ea9291e50922d976b2daf52feaef7c57R773
I also left some checking in ErrorAttachments static methods, should i remove them? I think that it is convenient for the user to know right away, that he forgot to provide required arguments!

Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java]: He will know as you validate right after. And either way its when you look at the logs that you notice that. Plus you will have trouble with code coverage if you validate something in double.

File: sdk/mobile-center-crashes/src/androidTest/java/com/microsoft/azure/mobile/crashes/CrashesAndroidTest.java [hunk 6]
```diff
@@ -172,6 +172,7 @@ public void onResult(ErrorReport errorReport) {
         assertNotNull(log.get());
         assertEquals(1, ErrorLogHelper.getErrorStorageDirectory().listFiles().length);
 
+        /* TODO (getErrorAttachment): Re-enable error attachment when the feature becomes available. */
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/androidTest/java/com/microsoft/azure/mobile/crashes/CrashesAndroidTest.java]: what is that TODO?

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/AbstractCrashesListener.java [hunk 7]
```diff
@@ -17,6 +17,7 @@ public boolean shouldAwaitUserConfirmation() {
         return false;
     }
 
+    /* TODO (getErrorAttachment): Re-enable error attachment when the feature becomes available. */
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/AbstractCrashesListener.java]: what is that TODO?

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 8]
```diff
@@ -738,6 +739,8 @@ public void run() {
                         Map.Entry<UUID, ErrorLogReport> unprocessedEntry = unprocessedIterator.next();
                         ErrorLogReport errorLogReport = unprocessedEntry.getValue();
                         mChannel.enqueue(errorLogReport.log, ERROR_GROUP);
+
+                        /* TODO (getErrorAttachment): Re-enable error attachment when the feature becomes available. */
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: what is that TODO? You restored the feature.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/CrashesListener.java [hunk 9]
```diff
@@ -31,12 +31,14 @@
...
     /**
-     * Called to get additional information to be send as separate ErrorAttachmentLog logs
+     * Called to get additional information to be sent as separate ErrorAttachmentLog logs
      * Attachments are optional so this method can also return <code>null</code>.
      *
      * @param report The crash report for additional information.
      * @return {@link Iterable<ErrorAttachmentLog>} instances of ErrorAttachmentLog to be sent as separate logs.
      */
+
+    /* TODO (getErrorAttachment): Re-enable error attachment when the feature becomes available. */
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/CrashesListener.java]: what is that TODO now that the feature is restored?

File: sdk/mobile-center-crashes/src/test/java/com/microsoft/azure/mobile/crashes/CrashesTest.java [hunk 10]
```diff
@@ -301,13 +301,15 @@ public void queuePendingCrashesShouldProcess() throws IOException, ClassNotFound
         when(mockListener.shouldProcess(report)).thenReturn(true);
         when(mockListener.shouldAwaitUserConfirmation()).thenReturn(false);
 
+        /* TODO (getErrorAttachment): Re-enable error attachment when the feature becomes available. */
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/test/java/com/microsoft/azure/mobile/crashes/CrashesTest.java]: TODO?

File: sdk/mobile-center-crashes/src/test/java/com/microsoft/azure/mobile/crashes/CrashesTest.java [hunk 11]
```diff
@@ -301,13 +301,15 @@ public void queuePendingCrashesShouldProcess() throws IOException, ClassNotFound
...
         ErrorAttachmentLog mockAttachment = mock(ErrorAttachmentLog.class);
         when(mockAttachment.getId()).thenReturn(UUID.randomUUID());
         when(mockAttachment.getErrorId()).thenReturn(UUID.randomUUID());
         when(mockAttachment.getContentType()).thenReturn("");
         when(mockAttachment.getFileName()).thenReturn("");
         when(mockAttachment.getData()).thenReturn("");
 
+        /* TODO (getErrorAttachment): Re-enable error attachment when the feature becomes available. */
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/test/java/com/microsoft/azure/mobile/crashes/CrashesTest.java]: TODO?

**Comment by jaelim-ms** [comment]
What I meant was that @bmourat need to keep the TODO comments as well as the lines that were commented out. The backend isn't ready so you shouldn't enable it unless the PR merges after backend is ready.
@guperrot: For the one that you talked offline, we should create another branch and the merge this PR to the new branch if that's the case. `don't merge` label doesn't say that we will wait until the backend is ready. You could have mentioned the intention of the label in the comment.

**Comment by guperrot** [comment]
@bmourat Ok it's settled then, please remove the TODO comments, sorry for that. We'll make sure the PR is not merged to develop until backend ready, no need for a new branch for a single PR.

**Comment by guperrot** [comment]
Created a new target branch so that we can have the future approved code on the main repos and build internal releases from there.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java [hunk 12]
```diff
@@ -22,52 +25,75 @@
...
+    @VisibleForTesting
+    static boolean isBinaryContentType(String contentType) {
+        return contentType != null && !contentType.startsWith("text/");
+    }
+
+    /**
+     * Validates ErrorAttachmentLog
+     *
+     * @param log ErrorAttachmentLog to validate.
+     * @return True if validation succeeded, otherwise False.
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ErrorAttachments.java]: We don't use that convention for booleans on Javadoc, we use `<code>true</code>` and `<code>false</code>`.

**Pull request merged**